policy.go

```diff
@@ -508,28 +488,75 @@ func computeStaticPolicy(alg tpm2.HashAlgorithmId, input *staticPolicyComputePar
 // computeDynamicPolicy computes the part of an authorization policy associated with a sealed key object that can change and be
 // updated.
 func computeDynamicPolicy(alg tpm2.HashAlgorithmId, input *dynamicPolicyComputeParams) (*dynamicPolicyData, error) {
-	var pcrData []policyPCRData
+	// Build a PCRSelectionList from input.pcrParams
+	var pcrs tpm2.PCRSelectionList
 	for _, p := range input.pcrParams {
 		if len(p.digests) == 0 {
-			return nil, fmt.Errorf("no digests provided for PCR%d", p.pcr)
+			return nil, fmt.Errorf("no digests provided for PCR%d, bank %v", p.pcr, p.alg)
+		}
+		var s *tpm2.PCRSelection
+		for i := range pcrs {
+			if pcrs[i].Hash == p.alg {
+				s = &pcrs[i]
+				break
+			}
+		}
+		if s == nil {
+			pcrs = append(pcrs, tpm2.PCRSelection{Hash: p.alg})
+			s = &pcrs[len(pcrs)-1]
+		}
+		for _, i := range s.Select {
+			if i == p.pcr {
+				return nil, fmt.Errorf("digests for PCR%d, bank %v provided more than once", p.pcr, p.alg)
+			}
+		}
+		s.Select = append(s.Select, p.pcr)
+	}
+
+	// Take every combination of digest from input.pcrParams and compute the policy digest that would result from the TPM2_PolicyPCR
+	// assertion with each combination
+	var pcrOrDigests tpm2.DigestList
+	// Keep track of the index in to each policyPcrParam.digests slice without having to do a recursive iteration.
+	indices := make([]int, len(input.pcrParams))
+Outer:
+	for {
+		// Loop over input.pcrParams and add the digest at the current index for each one to pcrValues
+		pcrValues := make(tpm2.PCRValues)
+		for i := range input.pcrParams {
+			pcrValues.SetValue(input.pcrParams[i].pcr, input.pcrParams[i].alg, input.pcrParams[i].digests[indices[i]])
+		}
+
+		// Compute PCR digest from the map of PCR values and the computed selection
+		digest, _ := tpm2.ComputePCRDigest(alg, pcrs, pcrValues)
+
+		// Execute trial TPM2_PolicyPCR with the computed PCR digest and selection, and save the result
+		trial, _ := tpm2.ComputeAuthPolicy(alg)
+		trial.PolicyPCR(digest, pcrs)
+		pcrOrDigests = append(pcrOrDigests, trial.GetDigest())
+
+		// Advance to the next digest combination
+		if len(input.pcrParams) == 0 {
+			// Special case - there are no PCR parameters so we've created a single digest for a TPM2_PolicyPCR assertion with
+			// an empty selection.
+			break
 		}
-		var orDigests tpm2.DigestList
-		for _, d := range p.digests {
-			trial, _ := tpm2.ComputeAuthPolicy(alg)
-			if len(pcrData) > 0 {
-				trial.PolicyOR(ensureSufficientORDigests(pcrData[len(pcrData)-1].OrDigests))
+		for i := range input.pcrParams {
+			indices[i]++
+			if indices[i] < len(input.pcrParams[i].digests) {
+				break
+			}
+			// The index for this policyPcrParam has advanced beyond the end of its digests slice - roll over to zero before moving on to
+			// advance the index for the next policyPcrParam
+			indices[i] = 0
+			if i == len(input.pcrParams)-1 {
+				// We've rolled all of the indices back to zero, so we're finished now
+				break Outer
```

Comment:
(nitpick) I wonder if this would be easier to read if this would get extracted into a helper and then it could return here instead of breaking out to "outer". But just a hunch

Reply:
I've updated it to use a helper function now.